src/azul/plugins/repository/tdr/__init__.py

```diff
@@ -156,50 +169,59 @@ def _choose_one_version(self, versioned_items: BigQueryRows) -> BigQueryRow:
-        bundle_project_id = links_row['project_id']
-        log.info('Retrieved links content, %s top-level links', len(links_json['links']))
-        bundle.add_entity('links.json', 'links', links_row)
-
-        entities = defaultdict(set)
-        entities['project'].add(bundle_project_id)
-        for link in links_json['links']:
-            link_type = link['link_type']
-            if link_type == 'process_link':
-                entities[link['process_type']].add(link['process_id'])
-                for category in ('input', 'output', 'protocol'):
-                    for entity_ref in link[category + 's']:
-                        entities[entity_ref[category + '_type']].add(entity_ref[category + '_id'])
-            elif link_type == 'supplementary_file_link':
-                # For MVP, only project entities can have associated supplementary files.
-                associate_type = link['entity']['entity_type']
-                associate_id = link['entity']['entity_id']
-                require(associate_type == 'project',
-                        f'Supplementary file must be associated with entity of type "project", '
-                        f'not "{associate_type}"')
-                require(associate_id == bundle_project_id,
-                        f'Supplementary file must be associated with the current project '
-                        f'({bundle_project_id}, not {associate_id})')
-                for supp_file in link['files']:
-                    entities['supplementary_file'].add(supp_file['file_id'])
-            else:
-                raise RequirementError(f'Unexpected link_type: {link_type}')
-
-        def retrieve_rows(entity_type: str, entity_ids: Set[str]):
+        def retrieve_links(bundle: TDRBundle, links_id: BundleFQID) -> Tuple[JSON, EntityID]:
+            links_columns = ', '.join(bundle.metadata_columns | {'content', 'project_id', 'links_id'})
+            links_row = one(self._run_sql(f'''
+                SELECT {links_columns}
+                FROM {self._source.bq_name}.links
+                WHERE links_id = '{links_id.uuid}'
+                    AND version = TIMESTAMP('{links_id.version}')
+            '''))
+            links_json = json.loads(links_row['content'])
+            bundle_project_id = links_row['project_id']
+            log.info('Retrieved links content, %s top-level links', len(links_json['links']))
+            bundle.add_entity('links.json', 'links', links_row)
+            return links_json, bundle_project_id
+
+        entities: TypeGroupedEntities = {}
+        unprocessed_fqids = {root_bundle_fqid}
+        processed_fqids = set()
+        # Recursively follow dangling inputs to collect entities from upstream
+        # bundles. Assumes subgraphs connected by dangling edges form a DAG; no
+        # subgraph will be processed more than once, but there is no defense
+        # against cycles.
+        while unprocessed_fqids:
+            # This ought to be concurrent
```

Comment:
I don't think so, considering that we're already tripping BQ limits.

https://github.com/DataBiosphere/azul/issues/2281